### ISSUE TITLE
Shared dictionary compression support 

### ISSUE THREAD

**Issue by nickchomey** [caddyserver/caddy#6204]
Shared Dictionary Compression is a new technique that is starting to become available in browsers, which can reduce data transfer by up to 98%! It does this by leveraging brotli (not relevant to caddy) and ztsd's ability to use custom compression dictionaries when compressing a file - using, for example, the previous version of a script to compress the new version and sending the difference. 

More on it all here https://developer.chrome.com/blog/shared-dictionary-compression

And here https://www.debugbear.com/blog/shared-compression-dictionaries

It is still in an Origin Trial in Chrome until April 30, 2024, so perhaps the Caddy team could inquire about it such that support could be implemented when appropriate? 

The zstd package used by caddy already supports custom dictionaries https://github.com/klauspost/compress/blob/de4073a3abdd00a2a95e608f9fcaf6ebf9141cc0/zstd/README.md?plain=1#L324

So, it would be a matter of adding and listening for the relevant headers to and from the browser client. 

Thanks for the consideration!

**Comment by mholt** [comment]
We could do that. (I'm curious how they solved the privacy problems associated with shared dictionaries. I haven't read up on it.)

Doesn't surprise me that zstd supports it, but that's the least common encoding browsers accept, currently. We'd need support in gzip and brotli libs as well.

Will keep an eye on it!

**Comment by ottenhoff** [comment]
This is exciting especially for those of us that pay egress bills to the cloud providers.

The custom dictionary is public with no ability for authentication. Ensuring privacy of information inside the dictionary is completely up to the implementor. [Command-line training](https://github.com/facebook/zstd/blob/dev/programs/zstd.1.md#dictionary-builder) on a few dozen files looks like the way most would implement it.

**Comment by ottenhoff** [comment]
> The main point of shared dictionaries is that there's no singular custom dictionary. Literally every file/script can be its own custom dictionary, allowing up to like 99% compression on file changes.

Agreed, but implementing this in a load balancer means holding state of former files and that sounds like a lot of work. Implementing one single dictionary as an argument to Caddy's existing "encode zstd" sounds relatively straightforward?

**Comment by mholt** [comment]
An option to load a dictionary file should be pretty simple, as mentioned above. I'd welcome a PR to review!

**Comment by shashank-priyadarshi** [comment]
@mholt can I please work on this?

**Comment by mholt** [comment]
@shashank-priyadarshi What is your plan for implementation?

**Comment by Zetanova** [comment]
Lets say dictionary compression is nice.

The minimum and missing part that caddy would need to provide beside the compression of the content itself,
is an extension of the https://caddyserver.com/docs/caddyfile/directives/encode directive to be able to set
to be used dictionary file (or maybe remote uri) for each encoding (dcb and/or dcz)

The reason being that only the dictionary file is required, 
is that the rest of the flow could fully provided by the current caddy features
and the dictionary training off-loded to an external process.

The external process (not by caddy) could be provided or discover 
match pattern of a website and fetch some samples to train the dictionary on.

After it could generate a caddy file to include and upload it to caddy and also the dictionaries. 

Example config to include would be then:
```caddy

# Map the sha256 to the human readable dictionary-id
map {http.request.header.Available-Dictionary} {dict_id} {
    "sha256-abc123="      "news-123"
    "sha256-def456="      "products-456"
    "sha256-ghi789="      "framework-789"
    default ""
}

vars {
    dict_identifier {http.request.header.Dictionary-ID}
    dict_identifier {vars.dict_identifier}{?vars.dict_id}
}

# When serving a .dict file, add the Use-As-Dictionary header
handle /dicts/*.dict {
    file_server
    header /dicts/framework-123.* Use-As-Dictionary "match=\"/static/*\", id=\"framework-123\""
    header /dicts/news-123.* Use-As-Dictionary "match=\"/news/*\", id=\"news-123\""
    header /dicts/products-123.* Use-As-Dictionary "match=\"/products/*\", id=\"products-123\""    
}

encode {?vars.dict_identifier} {
    dictionary { 
         zstd /dicts/{vars.dict_identifier}.zstd.dict
         br /dicts/{vars.dict_identifier}.br.dict
    }
    # Implicit fallback: to standard zstd/br/gzip compression if dictionary compression fails or is not applicable.
}
```

The dictionary file itself could also be fetched from an upstream server and cached,
this would simplify the upload of it to a caddy server with read-only or ephemeral filesystem.

The config could also be injected over the caddy-api and if the dictionary file can be fetched 
then no upload to caddy would be required. 

The dictionary file endpoint could support a fetch on both dictionary-id and available-dictionary,
then the map could also be removed.
